### PR TITLE
build, core, unix, test: add Haiku support to libuv

### PR DESCRIPTION
**Pull request by alaviss**
This PR introduce support for the [Haiku operating system](https://haiku-os.org). It's currently developed against Haiku's latest nightly release (can be downloaded [here](https://download.haiku-os.org/)).

Not all tests are passing at the moment, and this is and will be a fairly dirty branch. I'm working on tests and (if possible), fs events support. This branch will be rebased against master and marked as ready to review once I'm done with improving the test results.

Currently about 291/351 tests are passing (incl. fs events tests).

### CODE REVIEW

**Comment by bnoordhuis** [comment]
Haven't done an in-depth review yet but - since your changes don't seem to change the API or ABI on other platforms - you could target the v1.x branch, that's what we use for releases.

File: src/unix/haiku.c [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+  if (buffer == NULL || size == NULL || *size == 0)
+    return UV_EINVAL;
+
+  status = find_path(B_APP_IMAGE_SYMBOL, B_FIND_PATH_IMAGE_PATH, NULL, abspath,
+                     sizeof(abspath));
+  if (status != B_OK)
+    return UV__ERR(status);
+
+  abspath_len = strlcpy(buffer, abspath, *size);
```
Comment (bnoordhuis) [inline src/unix/haiku.c]: Just a note: use `uv__strscpy()` unless there's good reason to prefer strlcpy() here - but in that case please add an explaining comment.

File: src/unix/haiku.c [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  uv_cpu_info_t* cpu_info;
+
+  if (cpu_infos == NULL || count == NULL)
+    return UV_EINVAL;
+
+  status = get_cpu_topology_info(NULL, &topology_count);
+  if (status != B_OK)
+    return UV__ERR(status);
+
+  topology_infos = uv__malloc(topology_count * sizeof(cpu_topology_node_info));
```
Comment (bnoordhuis) [inline src/unix/haiku.c]: Style: libuv uses `sizeof(*topology_infos)`

File: src/unix/haiku.c [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+  status = get_system_info(&system);
+  if (status != B_OK) {
+    return UV__ERR(status);
+  }
+
+  *cpu_infos = uv__calloc(system.cpu_count, sizeof(**cpu_infos));
+  if (*cpu_infos == NULL) {
+    return UV_ENOMEM;
+  }
```
Comment (bnoordhuis) [inline src/unix/haiku.c]: Style nit: no braces for single-line consequents (ditto line 169.)

File: src/unix/core.c [hunk 4]
```diff
@@ -1348,6 +1348,7 @@ int uv_cpumask_size(void) {
 #endif
 }
 
+
```
Comment (cjihrig) [inline src/unix/core.c]: Tiny nit, but can you remove this extra line.

File: src/unix/haiku.c [hunk 5]
```diff
@@ -0,0 +1,174 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (cjihrig) [inline src/unix/haiku.c]: Can you use this copyright instead: https://github.com/libuv/libuv/blob/ce7f9c889e2afb7791f6385471abc9032ac91ff5/src/unix/ibmi.c#L1-L20

File: src/unix/pipe.c [hunk 6]
```diff
@@ -52,7 +52,7 @@ int uv_pipe_bind(uv_pipe_t* handle, const char* name) {
   name_len = strlen(name);
 
   if (name_len > sizeof(saddr.sun_path) - 1)
-    return -ENAMETOOLONG;
+    return UV_ENAMETOOLONG;
```
Comment (cjihrig) [inline src/unix/pipe.c]: This should probably be a separate commit/PR since it's not really related to Haiku support.

File: src/unix/haiku.c [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    return UV_ENOMEM;
+
+  status = get_cpu_topology_info(topology_infos, &topology_count);
+  if (status != B_OK) {
+    uv__free(topology_infos);
+    return UV__ERR(status);
+  }
+
+  cpuspeed = 0;
+  for (i = 0; i < (int)topology_count; i++)
```
Comment (cjihrig) [inline src/unix/haiku.c]: Please add braces around the loop body.

File: src/unix/haiku.c [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+  if (status != B_OK) {
+    return UV__ERR(status);
+  }
+
+  *cpu_infos = uv__calloc(system.cpu_count, sizeof(**cpu_infos));
+  if (*cpu_infos == NULL) {
+    return UV_ENOMEM;
+  }
+
+  /* CPU time is not exposed by Haiku, neither does the model name. */
```
Comment (cjihrig) [inline src/unix/haiku.c]: "does" -> "is" ... or reword as something like: "CPU time and model are not exposed by Haiku."

**Comment by cjihrig** [comment]
@alaviss are you volunteering to help maintain Haiku support over time?

**Comment by alaviss** [comment]
> @alaviss are you volunteering to help maintain Haiku support over time?

Sure, and @return has also expressed interest on doing this as well :)

**Comment by alaviss** [comment]
The implementation has since been imported to HaikuPorts, and @return have got nodejs and npm performed rather well on Haiku, so I'm marking this as ready for review.

There are still a lot of test failures (about 59 of them), but I'm attributing most of them as Haiku bugs and we'll solve them there. Further improvements will be sent in future PRs if any.

**Comment by bnoordhuis** [comment]
Landed in abe4f3d5, thanks! I took the liberty of squashing it down to a single commit. Thanks again!
